### PR TITLE
Migrate contracts to ink4

### PR DESCRIPTION
**Pull request by deuszx**
Currently targeting feature branch since our Testnet is not yet compatible with ink4 beta.

### CODE REVIEW

File: contracts/highlighted_posts/Cargo.toml [hunk 1]
```diff
@@ -5,15 +5,10 @@ authors = ["[your_name] <[your_email]>"]
 edition = "2021"
 
 [dependencies]
-ink_primitives = { version = "~3.4.0", default-features = false }
-ink_metadata = { version = "~3.4.0", default-features = false, features = ["derive"], optional = true }
-ink_env = { version = "~3.4.0", default-features = false }
-ink_storage = { version = "~3.4.0", default-features = false }
-ink_lang = { version = "~3.4.0", default-features = false }
-ink_prelude = { version = "~3.4.0", default-features = false }
+ink = { git = "https://github.com/paritytech/ink", rev = "4655a8b4413cb50cbc38d1b7c173ad426ab06cde", default-features = false}
```
Comment (h4nsu) [inline contracts/highlighted_posts/Cargo.toml:8]: Could you please add the comment with OpenBrush reference here too?

File: contracts/bulletin_board/lib.rs [hunk 2]
```diff
@@ -31,26 +31,24 @@
...
-    use ink_storage::{
-        traits::{PackedLayout, SpreadAllocate, SpreadLayout},
-        Mapping,
+    use ink::{
+        codegen::EmitEvent,
+        prelude::{format, string::String, vec::Vec},
+        reflect::ContractEventBase,
+        storage::{traits::StorageLayout, Mapping},
+        ToAccountId,
     };
```
Comment (h4nsu) [inline contracts/bulletin_board/lib.rs:52]: nit: combine these?

**Review by h4nsu — APPROVED** [review]

File: contracts/bulletin_board/lib.rs [hunk 3]
```diff
@@ -90,23 +88,14 @@ mod bulletin_board {
...
-        Eq,
-        scale::Encode,
-        scale::Decode,
-        SpreadLayout,
-        PackedLayout,
-    )]
-    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    // We will be storing `Bulletin` as _value_ in our contract (value type in
+    // `Mapping<K, Value>`) and for that reason we need to derive
+    // `ink::storage::traits::StorageLayout` for `std` feature.
```
Comment (h4nsu) [inline contracts/bulletin_board/lib.rs:93]: btw: do you know what exactly `StorageLayout` is responsible for and why do we need in only in `std` run?

File: contracts/bulletin_board/lib.rs (hunk 2 above)
Comment (deuszx) [reply contracts/bulletin_board/lib.rs:52]: Ah, poor rustfmt can't apply formatting inside a macro?

File: contracts/bulletin_board/lib.rs (hunk 3 above)
Comment (deuszx) [reply contracts/bulletin_board/lib.rs:93]: The only place where I could find it in the expanded macro was in `fn __ink_generate_metadata() -> ink::metadata::InkProject`, where it's used to:
* [`validate`](https://paritytech.github.io/ink/src/ink_metadata/layout/validate.rs.html#33-39)
* generate the metadata

**Pull request merged**